src/metabase/notification/send.clj

```diff
@@ -188,12 +181,89 @@
     nil))
 
+(defprotocol NotificationQueueProtocol
+  "Protocol for notification queue implementations."
+  (put-notification!  [this notification] "Add a notification to the queue. If a notification with the same id is already in the queue, replace it.")
+  (take-notification! [this]              "Take the next notification from the queue, blocking if none available."))
+
+(deftype ^:private NotificationQueue
+         [^java.util.LinkedList ids-list
+          ^java.util.concurrent.ConcurrentHashMap id->notification
+          ^java.util.concurrent.locks.ReentrantLock queue-lock
+          ^java.util.concurrent.locks.Condition not-empty-cond]
+  NotificationQueueProtocol
+  (put-notification! [_ notification]
+    (let [id (or (:id notification) (str (random-uuid)))]
+      (.lock queue-lock)
+      (try
+        (when-not (.contains ids-list id)
+          (.add ids-list id))
+        (.put id->notification id notification)
+        ;; Signal that a notification is available
+        (.signal not-empty-cond)
+        (finally
+          (.unlock queue-lock)))))
+
+  (take-notification! [_]
+    (.lock queue-lock)
+    (try
+      ;; Wait until there's at least one notification
+      (while (.isEmpty ids-list)
+        (.await not-empty-cond))
+      (let [id           (.removeFirst ids-list)
+            notification (.remove id->notification id)]
+        notification)
+      (finally
+        (.unlock queue-lock)))))
+
+(defn- create-notification-queue
+  "A thread-safe, notification queue with the following properties:
+  - Notifications are identified by unique IDs
+  - Adding a notification with an ID already in the queue replaces the existing one
+  - Taking from an empty queue blocks until a notification is available
+  - Multiple threads can safely add and take from the queue concurrently"
+  []
+  (let [queue-lock     (java.util.concurrent.locks.ReentrantLock.)
+        not-empty-cond (.newCondition queue-lock)]
+    (->NotificationQueue (java.util.LinkedList.)
+                         (java.util.concurrent.ConcurrentHashMap.)
+                         queue-lock
+                         not-empty-cond)))
+
+(defn- create-notification-dispatcher
+  "Create a thread pool for sending notifications.
+  There can only be one notification with the same id in the queue.
+  - if a notification of the same id is already in the queue, then replace it
+  - if a notification doesn't have id, put it into queue regardless (used to send unsaved notifications)"
+  [pool-size queue]
+  (let [executor (Executors/newFixedThreadPool
+                  pool-size
+                  (.build
+                   (doto (BasicThreadFactory$Builder.)
+                     (.namingPattern "send-notification-thread-pool-%d"))))]
+    (.addShutdownHook
+     (Runtime/getRuntime)
+     (Thread. ^Runnable (fn []
+                          (.shutdown ^ExecutorService executor))))
```

Comment:
`.shutdown` only stops new tasks being submitted. You probably want `.shutdownNow` which will `.interrupt` the task threads. You will also need to handle interrupts in the loop.